feat: Add PopoverNext component built on Floating UI

ggdouglas opened the pull request:
### Summary

Introduce `PopoverNext` as a modern, Floating UI–based replacement for `Popover`. The goal is near drop-in API compatibility to ease adoption, while migrating from Popper ["modifiers"](https://popper.js.org/docs/v2/modifiers) to Floating UI ["middleware"](https://floating-ui.com/docs/middleware). This change also helps to unblock React 19 support as part of #7115.

- **Why**: Popper v2 via `react-popper` isn’t actively maintained, and the ecosystem has moved to Floating UI. React 19 support is blocked by the legacy stack.
- **What**: New `PopoverNext` component, docs, tests, and examples. Existing `Popover` component marked deprecated.
- **How**:
  - Ported the existing Popover from a class-based to a functional component, preserving existing behavior where possible.
  - Maintained feature parity while switching the positioning engine to Floating UI.
  - Exposed `middleware` as options instead of `modifiers` in new `PopoverNext` API.

<img width="3456" height="1670" alt="Screenshot 2025-09-18 at 13 42 19@2x" src="https://github.com/user-attachments/assets/975be468-13e2-490e-b872-14f04777f5e0" />

### Migration
- **Prop compatibility**
  - Preserved where feasible: `children`, `renderTarget`, `content`, `isOpen`/`defaultIsOpen`, `onInteraction`, `onClose`, `usePortal`, `hasBackdrop`, `interactionKind`, `minimal`, `matchTargetWidth`, etc.
- **Auto behavior**
  - The `"auto"` placement options are not supported in `PopoverNext`.
  - Instead, for automatic placement behavior, omit `placement` entirely. When `placement` is undefined, `PopoverNext` uses Floating UI's automatic placement by default (via `autoPlacement`). See the docs: [Floating UI autoPlacement](https://floating-ui.com/docs/autoPlacement).
- **Defaults and behavior notes**
  - Boundary defaults differ:
    - `Popover`: `boundary="clippingParents"`
    - `PopoverNext`: `boundary="clippingAncestors"`
    - To match legacy behavior precisely, set `boundary="clippingParents"` on `PopoverNext`.
  - Dismiss and accessibility are preserved:
    - `Classes.POPOVER_DISMISS`, `Classes.POPOVER_DISMISS_OVERRIDE`, `captureDismiss`
    - `aria-haspopup`/`aria-expanded`, backdrop + focus behavior via `Overlay2`
- **Modifiers → Middleware mapping**
  - `flip` → `flip`
  - `preventOverflow` → `shift` (use `boundary`/`rootBoundary`)
  - `offset` → `offset` (`mainAxis`, `crossAxis`)
  - `arrow` → `arrow`
- **Practical migration steps**
  - Switch imports to `PopoverNext`.
  - Use the same props; modify `placement`, `boundary`, etc. where needed
  - Translate any Popper modifiers to Floating UI middleware (migration utilities forthcoming)

### Follow-ups
- Migrate internal usages (`Tooltip`, menus, selects) to `PopoverNext`.
- Provide codemods and a migration wiki page as adoption ramps.
- Remove legacy `Popover` in a future major version

### References
- Origin issue: #7115
- #7139
- Popper v2: [docs](https://popper.js.org/docs/v2), [react-popper](https://popper.js.org/react-popper)
- Floating UI: [site](https://floating-ui.com), [migration guide](https://floating-ui.com/docs/migration#migration)

ggdouglas committed:
Install `@floating-ui/react` dependency in core

ggdouglas committed:
Create PopoverNext component as new Popover implementation

ggdouglas committed:
Write docs for PopoverNext and add examples

svc-palantir-github commented:
<h3>Write docs for PopoverNext and add examples</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/431fd517-18b6-4ab7-a5fa-f180ee050f47/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/431fd517-18b6-4ab7-a5fa-f180ee050f47/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/431fd517-18b6-4ab7-a5fa-f180ee050f47/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/431fd517-18b6-4ab7-a5fa-f180ee050f47/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>Write docs for PopoverNext and add examples</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/66ff6b84-6772-41c3-9073-fc36261dcbab/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/66ff6b84-6772-41c3-9073-fc36261dcbab/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/66ff6b84-6772-41c3-9073-fc36261dcbab/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/66ff6b84-6772-41c3-9073-fc36261dcbab/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>Write docs for PopoverNext and add examples</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/b5739b45-db3e-4782-b08f-e5722fd7e033/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/b5739b45-db3e-4782-b08f-e5722fd7e033/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/b5739b45-db3e-4782-b08f-e5722fd7e033/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/b5739b45-db3e-4782-b08f-e5722fd7e033/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>Write docs for PopoverNext and add examples</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/4a8d58ed-2d45-4a73-b724-f6f235758ff3/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/4a8d58ed-2d45-4a73-b724-f6f235758ff3/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/4a8d58ed-2d45-4a73-b724-f6f235758ff3/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/4a8d58ed-2d45-4a73-b724-f6f235758ff3/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

invliD commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,152 @@
+            ...floatingProps,
+            className: classNames(ownTargetProps.className, targetModifierClasses),
+            // if the consumer renders a tooltip target, it's their responsibility to disable that tooltip
+            // when *this* popover is open
```
This seems like an odd place for this comment.

ggdouglas replied:
Hm, this is an artifact from being ported over from the original component without being recontextualized. I can try to move it somewhere more appropriate. fce1b5d261e5019223d8526d6db40ba9a1d9b03b
https://github.com/palantir/blueprint/blob/d8783f00b3ea2a637b62d6afcd9890d41a8614e3/packages/core/src/components/popover/popover.tsx#L333-L334

invliD commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,152 @@
+        const clonedTarget = cloneElement(childTarget, {
+            ...childTargetProps,
+            className: classNames(childTarget.props.className, targetModifierClasses),
+            disabled: isOpen && isTooltipElement(childTarget) ? true : childTarget.props.disabled,
```
Let's just use a regular boolean expression instead of a ternary.
```suggestion
            disabled: (isOpen && isTooltipElement(childTarget)) || childTarget.props.disabled,
```

ggdouglas replied:
bef744acd3ada851303151d9a9e5aebb9f76ab6c

invliD commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,152 @@
+import type { PopoverNextProps } from "./popoverNextProps";
+import { type usePopover } from "./usePopover";
+
+interface PopoverTargetProps extends PopoverNextProps {
```
Might be cleaner to define the exact props this component takes instead of taking a superset of the entire popover.

mm-wang replied:
+1

ggdouglas replied:
Added a couple changes related to limiting the scope of these props: dbf5012f7a9c831d372feb467e03ab9cd3bb6d26 00801418ed415055b4595074d12bc9066d15e96f

invliD commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,152 @@
+        handleTargetBlur,
+        handleTargetContextMenu,
+        handleTargetFocus,
+        interactionKind = PopoverInteractionKind.CLICK,
```
Can we do this default on the popover level instead of the target?

ggdouglas replied:
Yep! 62ee536d382a7bd26cbd2a945df7bb1e7ee0b2c8

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+
+-   A more advanced API is available through the `renderTarget` prop. Here, PopoverNext provides you with all the
+    information necessary to render a functional popover with a [render prop](https://reactjs.org/docs/render-props.html).
+    You are responsible for then propogating that information with an
```
```suggestion
    You are responsible for then propagating that information with an
```

ggdouglas replied:
fce1b5d261e5019223d8526d6db40ba9a1d9b03b

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+    information necessary to render a functional popover with a [render prop](https://reactjs.org/docs/render-props.html).
+    You are responsible for then propogating that information with an
+    [object spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals)
+    to the `React.JSX.Element` returned from `renderTarget`.
```
This sentence is quite confusing. You're responsible for spreading information into a JSX.Element? You should spread it into the props, not the resulting React object. This also sounds like you _call_ renderTarget, and then you spread the data to the returned object, but you actually return the object after spreading the props.

ggdouglas replied:
Hm, yeah, another one that I copied over from the original Popover docs:
https://github.com/palantir/blueprint/blob/d8783f00b3ea2a637b62d6afcd9890d41a8614e3/packages/core/src/components/popover/popover.md?plain=1#L64-L66

I'll try to come up with some better wording here. fce1b5d261e5019223d8526d6db40ba9a1d9b03b

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+    to the `React.JSX.Element` returned from `renderTarget`.
+
+    -   If the rendered element is _not_ a native HTML element, you must ensure that it supports the
+        `className`, `ref`, and `tabIndex` props (i.e. renders them out to the DOM).
```
"into" maybe? Never heard "render out" before…
```suggestion
        `className`, `ref`, and `tabIndex` props (i.e. renders them into the DOM).
```

ggdouglas replied:
fce1b5d261e5019223d8526d6db40ba9a1d9b03b

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+
+@### Placement
+
+The `placement` prop controls the popover's position relative to the target. PopoverNext passes this prop directly to Floating UI; it uses the same semantics and supported values [as shown here in the docs](https://floating-ui.com/docs/computePosition#placement).
```
I think we should abstract this away. We can still refer to the Floating UI docs, but we should repeat the relevant parts (assuming the below is not complete) and remove "PopoverNext passes this prop directly to Floating UI"—that's implementation detail and may change in the future if they break API and we don't want to.

ggdouglas replied:
I can get on board with that; first crack at breaking some of this out into our own docs explicitly: fce1b5d261e5019223d8526d6db40ba9a1d9b03b

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+
+</div>
+
+@### Middleware
+
+Middleware allow us to customize Floating UI's positioning behavior. **PopoverNext** configures several of Floating UI's built-in middleware to handle things such as flipping, preventing overflow from a boundary element, and positioning the arrow.
+
+You may override the default middleware with the `middleware` prop, which is an object with key-value pairs representing the middleware name and its options object, respectively. See the [Floating UI middleware docs page](https://floating-ui.com/docs/middleware) for more info.
+
+<div class="@ns-callout @ns-intent-warning @ns-icon-warning-sign @ns-callout-has-body-content">
+    <h5 class="@ns-heading">Auto placement requires autoPlacement or flip middleware</h5>
+
+Be careful when disabling the "autoPlacement" or "flip" middleware, since automatic placement relies on them. If you _do_ decide to disable these middleware, be sure to also specify a placement which is specific (not automatic).
```
Do we need to support this API? It might be best to keep the simplest possible API for now until we know we need more, and even then maybe make it more semantic than just exposing all of Floating UI's API. That way, if we ever have to switch underlying libraries again, we might be able to get away without an API break next time.

ggdouglas replied:
I get the concern about keeping the API simple and future-proof, but I'm thinking we should probably keep the middleware prop for now.

The main thing is that we're going to need this flexibility pretty much immediately. Looking at the current codebase, submenus and tooltips already rely on custom modifiers for things like `offset`, `flip`, and `preventOverflow`.

https://github.com/palantir/blueprint/blob/031a998e066f234067e46ca37d4e5ad192c51b42/packages/core/src/components/menu/menuItem.tsx#L308-L314

https://github.com/palantir/blueprint/blob/031a998e066f234067e46ca37d4e5ad192c51b42/packages/core/src/components/tooltip/tooltip.tsx#L126-L135

Without exposing middleware, we'd either have to recreate all these specific APIs (which feels like more work for less flexibility) or continually add on to functionality during the migration.

I also think the middleware concept is actually pretty well-designed already. Floating UI's API is well-documented, and it's tough to anticipate all the positioning edge cases we might need to handle. A more semantic API would probably just end up being a thin wrapper around middleware anyway, and then we'd be maintaining our own abstraction layer.

On the future-proofing front, if we do need to switch libraries again someday, the middleware concept seems pretty universal. Popper had modifiers, Floating UI has middleware, so there's probably a pattern here. We could potentially create a compatibility layer that maps between different systems if needed.

What's your take? Are there specific middleware use cases that concern you, or do you have a suggestion on a clean wrapper we could use here?

invliD commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,369 @@
+@### Minimal style
+
+You can create a minimal popover by setting `minimal={true}`.
+This removes the arrow from the popover and makes the transitions more subtle.
```
I believe we've talked about this one before, but do we want to split these two into separate props now that we're redoing the API anyway?

ggdouglas replied:
Yeah, there are a few things to split out here:
- Presence of the arrow
- Offset of the popup element from the target
- Animation style

I've split `minimal` into two separate props, `arrow` and `animation`. The `arrow` prop controls the display of the arrow and applies the correct amount of offset to accommodate the arrow. The `animation` prop solely controls the style of the animation ("scale" vs. "minimal"). 86cd4c0b15ec6e6950d3d98a5b63e01e7672f7ac

invliD commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,72 @@
+import type { PopoverProps } from "../popover/popoverProps";
+import type { DefaultPopoverTargetHTMLProps } from "../popover/popoverSharedProps";
+
+export type { FloatingBoundary, FloatingPlacement };
```
We should probably export our own types that correspond to these. That would avoid accidentally propagating a breaking change if floating UI ever made one, and allows us to abstract away the use of the library at least in our API.

mm-wang replied:
If Floating UI made a breaking change, and we were to rely on the updated version, wouldn't we want to have that change included?

ggdouglas replied:
Turns out that `FloatingBoundary` no longer needed to be re-exported. Created our own type for Placement: 7dd3241d3238a319bde36653f2e9b091d15503d3

invliD replied:
> If Floating UI made a breaking change, and we were to rely on the updated version, wouldn't we want to have that change included?

Probably not, because that would require us to major bump Blueprint, and we'd likely not want to do that at the time prescribed by Floating UI. By making our own type, we can abstract that connection away, and have a conversion function that keeps our API stable, and we can then break ours at our convenience at the next Blueprint major bump.

invliD commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,72 @@
+ */
+export interface PopoverNextProps<T extends DefaultPopoverTargetHTMLProps = DefaultPopoverTargetHTMLProps>
+    extends Omit<PopoverProps<T>, "boundary" | "modifiers" | "placement" | "rootBoundary"> {
+    boundary?: FloatingBoundary;
```
Since this is public API we should probably redefine the actual types here. It'll make it easier to eventually delete Popover in favor of PopoverNext, and also makes this more readable.

ggdouglas replied:
I can get on board with this 👍  00801418ed415055b4595074d12bc9066d15e96f

invliD commented on packages/core/src/components/popover-next/popoverPopup.tsx:
```diff
@@ -0,0 +1,169 @@
+}
+
+function cssPropertyToString(value: string | number | undefined): string {
+    return value != null && !isNaN(Number(value)) ? `${value}px` : "";
```
What happens if you set `left: ` (the empty string case)? Should we set the property at all in that case?

mm-wang commented on packages/core/src/components/popover-next/floatingUtils.ts:
```diff
@@ -0,0 +1,67 @@
+        middlewares.push(autoPlacement(overrides.autoPlacement));
+    } else if (overrides?.autoPlacement) {
+        middlewares.push(autoPlacement(overrides.autoPlacement));
+    } else if (overrides?.flip) {
```
**nit/question**: I can tell that this is meant to be very explicit, but I was wondering if there's a reason why this isn't defaulting to pushing `autoPlacement(overrides.autoPlacement))` in an `else` block:
```
} else if (overrides?.flip) {
        middlewares.push(flip(overrides.flip));
} else {
        middlewares.push(autoPlacement(overrides.autoPlacement);
}
```
Is it because you might not have the overrides for placement at all? Mostly asking for my own understanding

ggdouglas replied:
Hmm, it's entirely possible that this got overcomplicated at some point in porting over the logic. Looking at where this is used in `PopoverNext`, it looks like it's always going to be provided with either `autoPlacement `or `flip` in its default config. Your suggestion makes sense to me. 06463f9ba453c4575024143003fdc4b528f87b92

mm-wang commented on packages/core/src/components/popover-next/floatingUtils.ts:
```diff
@@ -0,0 +1,67 @@
+ */
+export function convertMiddlewareConfigToArray(overrides: MiddlewareConfig | undefined): Middleware[] {
+    const middlewares: Middleware[] = [];
+
```
**nit/question**: Are there any other overrides than the ones below? If not, could you return early here? That might save checking if overrides exist below, but I'm asking more for my understanding.

ggdouglas replied:
Are you asking in regards to the need for optional chaining here? Looks like we can actually remove checks that `overrides` exists since we currently only use this utility function in one place where it's always provided. 9b3be1305dfcbd0a4e8d66097eeac863a6f9b853

Not sure if that helps clarify though. Let me know if I misunderstood!

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,423 @@
+
+export const PopoverNext = forwardRef<PopoverNextRef, PopoverNextProps>((props, ref) => {
+    const {
+        boundary = "clippingAncestors",
```
**note**: Thought it was odd that Floating UI didn't have an enum for this, but it looks like it's literally the string: https://floating-ui.com/docs/detectoverflow#boundary.

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,423 @@
+    const cancelOpenTimeout = useRef<(() => void) | undefined>(undefined);
+    const isMouseInTargetOrPopover = useRef(false);
+    const lostFocusOnSamePage = useRef(true);
+    const targetRef = createRef<HTMLElement>();
```
**question**: Why is this using `createRef` vs `useRef`?

ggdouglas replied:
Hm, good point! I didn't notice this when porting over logic from `Popover` but `useRef` is more appropriate here to maintain a stable ref. Changed in 290909173015874702c4b52666fbcffdd640de1b

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,423 @@
+    const isControlled = isOpen !== undefined;
+    const isContentEmpty = content == null || Utils.isEmptyString(content);
+
+    const computedIsOpen = disabled ? false : isOpen ?? defaultIsOpen;
```
**nit**: `const computedIsOpen = disabled ? false : (isOpen ?? defaultIsOpen)` for clarity, although I believe the latter evaluates first anyway, so really not a big deal

invliD replied:
Same as my comment above, this should probably just be `!disabled && (isOpen ?? defaultIsOpen)`. Ternaries with booleans in both arms just make the expression more difficult to follow.

ggdouglas replied:
Hm, I can agree on the point of clarity, but it seems that Prettier ends up stripping these parens out regardless.

invliD replied:
That's fine, I'm not too worried about the parens, just about a ternary with more boolean expressions in the two arms.

ggdouglas committed:
[CR] Edit comments/markdown

ggdouglas committed:
[CR] Simplify boolean expression in PopoverTarget

svc-palantir-github commented:
<h3>[CR] Edit comments/markdown</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/d76cf19a-373a-4bfb-8619-1f70a132f5a0/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/d76cf19a-373a-4bfb-8619-1f70a132f5a0/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/d76cf19a-373a-4bfb-8619-1f70a132f5a0/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/d76cf19a-373a-4bfb-8619-1f70a132f5a0/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
[CR] Handle empty strings in `cssPropertyToString`

ggdouglas committed:
[CR] Remove redundant `interactionKind` default in `PopoverTarget`

svc-palantir-github commented:
<h3>[CR] Remove redundant `interactionKind` default in `PopoverTarget`</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/27df7013-5321-425c-b9af-c008cf79064e/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/27df7013-5321-425c-b9af-c008cf79064e/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/27df7013-5321-425c-b9af-c008cf79064e/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/27df7013-5321-425c-b9af-c008cf79064e/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
[CR] Define exact props for PopoverTarget and PopoverPopup components

ggdouglas committed:
[CR] Explicitly redefine props for PopoverNext

ggdouglas committed:
[CR] Redeclare new Placement type within Blueprint

svc-palantir-github commented:
<h3>[CR] Redeclare new Placement type within Blueprint</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/675401f2-6077-49fe-a595-4a49827697ea/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/675401f2-6077-49fe-a595-4a49827697ea/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/675401f2-6077-49fe-a595-4a49827697ea/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/675401f2-6077-49fe-a595-4a49827697ea/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang commented on packages/core/src/components/popover-next/popoverPopup.tsx:
```diff
@@ -0,0 +1,169 @@
+        popoverClassName,
+    );
+
+    const defaultAutoFocus = isHoverInteractionKind ? false : undefined;
```
**question**: What is the difference between `false` and `undefined` here?

ggdouglas replied:
This logic was originally defined in the original `Popover` component here:
https://github.com/palantir/blueprint/blob/d8783f00b3ea2a637b62d6afcd9890d41a8614e3/packages/core/src/components/popover/popover.tsx#L415

My understanding of this is that for hover interactions, we want to explicitly disable auto-focus (`false`) because we don't want hover to steal focus from the user's current input. For click interactions, we want to use the default behavior which likely enables auto-focus when appropriate.

mm-wang reviewed (commented):
Reviewed for syntax and understanding, mostly tried to get context and an example of how we structure these components. I'd love to hear more about how I can use and test this.

ggdouglas committed:
[CR] Simplify middleware placement logic

ggdouglas committed:
Remove optional chaining from middleware config

svc-palantir-github commented:
<h3>[CR] Simplify middleware placement logic</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/4c5f0f4b-8738-4c62-8fc3-a50da71aa28c/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/4c5f0f4b-8738-4c62-8fc3-a50da71aa28c/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/4c5f0f4b-8738-4c62-8fc3-a50da71aa28c/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/4c5f0f4b-8738-4c62-8fc3-a50da71aa28c/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
[CR] Use `useRef` instead of `createRef` for targetRef in PopoverNext

Replace createRef with useRef for targetRef to maintain ref persistence
across re-renders

svc-palantir-github commented:
<h3>[CR] Use `useRef` instead of `createRef` for targetRef in PopoverNext</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/53d86660-0c21-4945-b3e8-a297524bd530/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/53d86660-0c21-4945-b3e8-a297524bd530/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/53d86660-0c21-4945-b3e8-a297524bd530/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/53d86660-0c21-4945-b3e8-a297524bd530/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Split `minimal` prop into new `arrow` and `animation` props

svc-palantir-github commented:
<h3>Split `minimal` prop into new `arrow` and `animation` props</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/53187f60-724d-4a7d-9126-6deeea767bf6/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/53187f60-724d-4a7d-9126-6deeea767bf6/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/53187f60-724d-4a7d-9126-6deeea767bf6/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/53187f60-724d-4a7d-9126-6deeea767bf6/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>Split `minimal` prop into new `arrow` and `animation` props</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/c699facd-a703-42e9-9731-1e8e1c08f082/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/c699facd-a703-42e9-9731-1e8e1c08f082/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/c699facd-a703-42e9-9731-1e8e1c08f082/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/c699facd-a703-42e9-9731-1e8e1c08f082/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Set `enforceFocus` to false in PopoverNext example

svc-palantir-github commented:
<h3>Set `enforceFocus` to false in PopoverNext example</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/4122a344-20f0-4c6b-b335-0028613186ea/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/4122a344-20f0-4c6b-b335-0028613186ea/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/4122a344-20f0-4c6b-b335-0028613186ea/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/4122a344-20f0-4c6b-b335-0028613186ea/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Merge branch 'develop' into gdouglas/popover-next

ggdouglas committed:
Run `pnpm format`

svc-palantir-github commented:
<h3>Merge branch 'develop' into gdouglas/popover-next</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/65b1cbfa-e857-437c-9bca-ae32c4d76f40/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/65b1cbfa-e857-437c-9bca-ae32c4d76f40/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/65b1cbfa-e857-437c-9bca-ae32c4d76f40/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/65b1cbfa-e857-437c-9bca-ae32c4d76f40/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>Run `pnpm format`</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/01eb4e09-083f-4d54-b2c6-75a658ccc866/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/01eb4e09-083f-4d54-b2c6-75a658ccc866/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/01eb4e09-083f-4d54-b2c6-75a658ccc866/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/01eb4e09-083f-4d54-b2c6-75a658ccc866/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Merge branch 'develop' into gdouglas/popover-next

ggdouglas committed:
Add boundary to PopoverNext example to match legacy Popover example

ggdouglas committed:
Add Blueprint-owned middleware types for PopoverNext

ggdouglas committed:
Ignore `jsx-a11y/no-autofocus` rule in PopoverPopup

svc-palantir-github commented:
<h3>Ignore `jsx-a11y/no-autofocus` rule in PopoverPopup</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/d617ace6-a48b-4bef-a0dc-cf0a37da7db9/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/d617ace6-a48b-4bef-a0dc-cf0a37da7db9/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/d617ace6-a48b-4bef-a0dc-cf0a37da7db9/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/d617ace6-a48b-4bef-a0dc-cf0a37da7db9/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

svc-palantir-github commented:
<h3>(temp) Add PopoverComparisonDemo</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/cfc11803-e617-4dd3-86ab-234421257571/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/cfc11803-e617-4dd3-86ab-234421257571/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/cfc11803-e617-4dd3-86ab-234421257571/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/cfc11803-e617-4dd3-86ab-234421257571/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang commented on packages/core/src/components/popover-next/popover-next.md:
```diff
@@ -0,0 +1,400 @@
+</h5>
+
+This component renders an **Overlay2** which works best inside a React tree which includes an
+**OverlaysProvider**. Blueprint v5.x includes a backwards-compatibile shim which allows this context
```
**question**: Why do we need Overlay2? Is it for interactions with other overlays?

mm-wang replied:
Interested to see porting this over in another implementation of this.

ggdouglas replied:
The primary reasons we use [Overlay2](https://blueprintjs.com/docs/#core/components/overlay2) in the original Popover component are:
- Portal functionality
- Focus capturing
- Rendering transitions
- (optionally) Rendering backdrop components

The current set of functionality with Overlay2 was preserved in PopoverNext while porting the component, but it is worth noting that Floating UI can (theoretically) supplant Overlay2 with the features it exposes within the library itself:
- https://floating-ui.com/docs/FloatingPortal
- https://floating-ui.com/docs/FloatingFocusManager
- https://floating-ui.com/docs/useTransition
- https://floating-ui.com/docs/FloatingOverlay

This was out-of-scope for the original intent behind this PR to craft PopoverNext as a drop-in replacement, but noting here for future reference.

mm-wang commented on packages/core/src/components/popover-next/usePopover.ts:
```diff
@@ -0,0 +1,99 @@
+        [onOpenChange, isControlled],
+    );
+
+    const data = useFloating({
```
**question**: Did we want to expose the floating strategy? https://floating-ui.com/docs/useFloating#strategy

ggdouglas replied:
Ah, good callout. I somehow missed the existing `positioningStrategy` prop on Popover when porting it to PopoverNext. Added in eda60938b0d421762937313a3e03c928f0d7984e for parity.

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+        position,
+        renderTarget,
+        rootBoundary,
+        shouldReturnFocusOnClose = false,
```
**suggestion**: I thought this was expected to be `true` to satisfy a11y requirements?
```suggestion
        shouldReturnFocusOnClose = true,
```

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+        placement,
+        position,
+        renderTarget,
+        rootBoundary,
```
**question**: Is this necessary, given that we have `boundary`? Could we instead extend boundary's capabilities? I would expect to try to minimize this API as opposed to expose everything.

mm-wang replied:
Sounds like we're already exposing this right now, so this doesn't help optimize.

ggdouglas replied:
We talked a bit offline re: this, but the basic BLUF here is:

- The current Popover component exposes both `boundary` and `rootBoundary`, so we aren't exposing any _additional_ props with PopoverNext
- To clarify the different between the two:
  - [`boundary`](https://floating-ui.com/docs/detectoverflow#boundary) defines local clipping containers (like a scrollable div)
  - [`rootBoundary`](https://floating-ui.com/docs/detectoverflow#rootboundary) defines the _ultimate_ boundary (viewport or entire document)
  - FloatingUI checks overflow against both and uses the most restrictive result

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+        usePortal,
+    ]);
+
+    const [hasDarkParent, setHasDarkParent] = useState(false);
```
**question**: Can we get rid of this? I'm not seeing usages unless I'm misreading this...

mm-wang replied:
Same thoughts as https://github.com/palantir/blueprint/pull/7573#discussion_r2747568860

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+        onInteraction,
+        openOnTargetFocus = true,
+        placement,
+        position,
```
**question**: What is the difference between `position` and `placement`? I thought position was determined by placement.

ggdouglas replied:
Hmm, good question. I originally kept both props here for component API parity, but I now ultimately think we can be a bit more opinionated and port just one for simplicity in the new component.

Here are the docs definitions:

https://github.com/palantir/blueprint/blob/90ce7125b971fef9926149c9f86ab16366db00ce/packages/core/src/components/popover/popoverSharedProps.ts#L256-L267

https://github.com/palantir/blueprint/blob/90ce7125b971fef9926149c9f86ab16366db00ce/packages/core/src/components/popover/popoverSharedProps.ts#L274-L284

Namely calling out this note on `placement`:
> Mutually exclusive with `position` prop. Prefer using this over `position`, as it more closely aligns with Popper.js semantics.

The rendered types look like the following:

For **`placement`**:
- `"top"`, `"top-start"`, `"top-end"`
- `"bottom"`, `"bottom-start"`, `"bottom-end"`
- `"left"`, `"left-start"`, `"left-end"`
- `"right"`, `"right-start"`, `"right-end"`
- `"auto"`, `"auto-start"`, `"auto-end"`

For **`position`**:
- `"top"`, `"top-left"`, `"top-right"`
- `"bottom"`, `"bottom-left"`, `"bottom-right"`
- `"left"`, `"left-top"`, `"left-bottom"`
- `"right"`, `"right-top"`, `"right-bottom"`
- `"auto"`, `"auto-start"`, `"auto-end"`

So `placement` is more of a [Popper.js](https://popper.js.org/docs/v2/constructors/#options) and [Floating UI](https://floating-ui.com/docs/computePosition#placement) concept that uses [logical operators](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Logical_properties_and_values) for each suffix of the side, whereas `position` is more of a Blueprint-derived concept.

Personally, I think we should opt to roll forward with just `placement` considering that it both:

1. Has better parity with Floating UI types
2. Aligns more generally with our movement towards properties with logical equivalents (see https://github.com/palantir/blueprint/issues/7193)

ggdouglas replied:
Removed `position` in 229f3df124068c631b17f5849bbe90dd9c9d80dc

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+                floatingData.update();
+            },
+        }),
+        [floatingData],
```
**suggestion**: Limit dependency to `floatingData.update`?
```suggestion
        [floatingData.update],
```

ggdouglas replied:
Seems like `react-hooks/exhaustive-deps` is not happy with that change:

<img width="1712" height="362" alt="Screenshot 2026-02-02 at 16 20 06@2x" src="https://github.com/user-attachments/assets/f459b0f4-2067-46fe-99b8-4ba9b6f09246" />

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+    }, [placement, boundary, rootBoundary, arrow, arrowRef, matchTargetWidth, middlewareOverrides]);
+
+    const floatingData = usePopover({
+        canEscapeKeyClose,
```
**question**: Should `positioningStrategy` be set here?

ggdouglas replied:
Added in eda60938b0d421762937313a3e03c928f0d7984e

mm-wang commented on packages/core/src/components/popover-next/popoverNext.tsx:
```diff
@@ -0,0 +1,415 @@
+
+    const updateDarkParent = useCallback(() => {
+        if (usePortal && floatingData.isOpen) {
+            setHasDarkParent(targetRef.current?.closest(`.${Classes.DARK}`) != null);
```
**question**: How are we planning to handle when we have the data attribute? This is assuming we still need `hasDarkParent`

ggdouglas replied:
I suppose we could update this logic to include the data attribute if/when we make a switch there. I imagine the implementation would look something like:
```ts
// Current: only checks for the class
setHasDarkParent(targetRef.current?.closest(`.${Classes.DARK}`) != null);

// Future: check for class OR data attribute
setHasDarkParent(
    targetRef.current?.closest(`.${Classes.DARK}, [data-theme="dark"]`) != null
);
```

mm-wang commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,434 @@
+     */
+    captureDismiss?: boolean;
+
+    /** Whether the popover can be closed by pressing the Escape key. */
```
**question**: Should we even allow this prop? Shouldn't we always expect that a user can use `Esc` to close a popover?

ggdouglas replied:
I'm not sure I have enough insight/context into why this was originally made an option on the original Popover component other than to say that it is originally exposed through `OverlayableProps` and is thus a feature of components that use Overlay in Blueprint: https://github.com/palantir/blueprint/blob/90ce7125b971fef9926149c9f86ab16366db00ce/packages/core/src/components/overlay/overlayProps.ts#L27-L32

Considering that this component is the base of many other, more specific components (e.g. Dialog, Tooltip, Select, etc.), I can see the value in having this be optional since there are contexts where you may/may not want to have the widget be closeable on Esc depending on the context. For that reason, I'm in favor of leaving this intact for now.

mm-wang replied:
Hmm... all right. I think if we default to true, that is fine, but I do find it odd to not allow it.

mm-wang commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,434 @@
+     *
+     * @default true
+     */
+    inheritDarkTheme?: boolean;
```
**suggestion**: This should not be a component prop, IMO. I think this should be done at the CSS level.

mm-wang replied:
In the spirit of trying for pure parity with `Popover`, fine to ignore this. Longer term would prefer to separate styling from the component API.

ggdouglas replied:
We talked a bit offline about the difficulties/nuance around `inheritDarkTheme`. Here's a summary:

The concern is that using JavaScript to control dark mode styling feels like it should be handled purely in CSS. However, there's a fundamental CSS limitation at play here: **CSS selectors can only traverse downward through the DOM tree.** When a popover is rendered into a portal (typically a sibling of the app root, not a descendant), it cannot inherit styles from `.bp6-dark` applied to an ancestor of the _trigger_ element through normal CSS cascade.

A potential CSS-only solution would be using `:root:has(.bp6-dark)` to apply dark styles globally, but this requires the dark mode class to be applied at or near the root level. Currently, Blueprint applications typically apply `.bp6-dark` to a container element wrapping their app, not at the document root.

Future solutions that could eliminate this prop:
- A `ThemeProvider` context that portaled components could consume
- Moving dark mode to a data-theme attribute on `:root` or `<body>`
- Global CSS custom properties for theming

These are larger architectural changes. For now, `inheritDarkTheme` maintains parity with the legacy Popover behavior and allows for easier migrations. We can revisit this as part of a broader theming initiative.

mm-wang commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,434 @@
+     * A root boundary element supplied to the positioning middleware.
+     * This is a shorthand for overriding Floating UI middleware options with the `middleware` prop.
+     */
+    rootBoundary?: PopoverNextRootBoundary;
```
**suggestion**: I don't know that I'd expose this.

mm-wang replied:
Looks like we already do so 🤷

ggdouglas replied:
See https://github.com/palantir/blueprint/pull/7573#discussion_r2756152951

mm-wang commented on packages/core/src/components/popover-next/popoverNextProps.ts:
```diff
@@ -0,0 +1,434 @@
+     * document after this popover closes.
+     *
+     * This is automatically set (overridden) to:
+     *  - `false` for hover interaction popovers
```
**question**: Why would this be false for tooltips? I understand that it should be true always: https://www.w3.org/WAI/ARIA/apg/patterns/tooltip/#:~:text=A%20tooltip%20is%20a%20popup%20that%20displays,dismissed%20when%20it%20no%20longer%20has%20focus

mm-wang commented on packages/core/src/components/popover-next/popoverPopup.tsx:
```diff
@@ -0,0 +1,162 @@
+
+    const defaultAutoFocus = isHoverInteractionKind ? false : undefined;
+
+    const shouldReturnFocusOnClose = isHoverInteractionKind
```
Noted my thoughts above.

mm-wang commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,140 @@
+/* !
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
```
**nit**: All 2026 now (sorry!)

ggdouglas replied:
Updated in cabca9ce3cebbe1cea99827d49b3d671f7bc3d3c

mm-wang reviewed (commented):
This is my first pass at review. I haven't taken a look at tests, examples, or styling. I also have only done a cursory review of popup + target, and probably still need to re-review. But here are some comments thus far.

I'd like to try to minimize the API boundary if possible - the current component has a pretty large surface area and this is increasing it with some added props. I'd love to know whether we can make choices for the user about what is actually useful to expose.

I also wanted to check - is there a way that we're detecting simulated clicks, similar to the old version?
```
isSimulatedButtonClick = (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) => {
        return !e.isTrusted && (e.target as HTMLElement).matches(`.${Classes.BUTTON}`);
    };
```

mm-wang commented:
I also want to check on a couple of things:
* Can we add the delays to open + close to the examples? That'd be good to note for a "kitchen sink" model
* Can we add a context menu opening another context menu to check that behavior with focus?

ggdouglas committed:
Update copyright year for new files

svc-palantir-github commented:
<h3>Update copyright year for new files</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/fa7e42e8-2a26-4e40-8a0c-03c011da691c/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/fa7e42e8-2a26-4e40-8a0c-03c011da691c/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/fa7e42e8-2a26-4e40-8a0c-03c011da691c/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/fa7e42e8-2a26-4e40-8a0c-03c011da691c/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Add `positioningStrategy` to PopoverNext

svc-palantir-github commented:
<h3>Add `positioningStrategy` to PopoverNext</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/69fa031e-2a4e-424d-8cb6-8418e4943972/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/69fa031e-2a4e-424d-8cb6-8418e4943972/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/69fa031e-2a4e-424d-8cb6-8418e4943972/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/69fa031e-2a4e-424d-8cb6-8418e4943972/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas commented:
> I also want to check on a couple of things:
> 
> * Can we add the delays to open + close to the examples? That'd be good to note for a "kitchen sink" model
> * Can we add a context menu opening another context menu to check that behavior with focus?

@mm-wang RE: these:

> Can we add the delays to open + close to the examples? That'd be good to note for a "kitchen sink" model

I attempted this locally, but I'm a bit wary of adding more examples to the "kitchen sink" here since it's getting to the point where there are just too many options / it's a bit overwhelming for the docs. It's also only really relevant when the interaction mode is a hover-type, so that makes these controls irrelevant otherwise.

<img width="811" height="986" alt="Screenshot 2026-02-02 at 17 57 57" src="https://github.com/user-attachments/assets/1b975eaa-35a7-452f-a2a2-f3b627c9ddb3" />

I think I want to defer on adding these, they seem like useful controls, but would maybe be better suited for inclusion in something like Storybook.

> Can we add a context menu opening another context menu to check that behavior with focus?

I'm a bit confused. Can you elaborate on where this should be added?

ContextMenu currently uses the legacy Popover, not PopoverNext. Adding that example would be useful for testing ContextMenu behavior in general, but it's out of scope for this PR. Happy to add it as a follow-up if you'd like, or we can revisit when/if ContextMenu is migrated to use PopoverNext.

ggdouglas committed:
Remove `position` prop from PopoverNext

svc-palantir-github commented:
<h3>Remove `position` prop from PopoverNext</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/9b4dcf25-01d4-451e-bc73-f873bac0b4df/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/9b4dcf25-01d4-451e-bc73-f873bac0b4df/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/9b4dcf25-01d4-451e-bc73-f873bac0b4df/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/9b4dcf25-01d4-451e-bc73-f873bac0b4df/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Modify the default behavior of `shouldReturnFocusOnClose` to be `true`

svc-palantir-github commented:
<h3>Modify the default behavior of `shouldReturnFocusOnClose` to be `true`</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/f66c4453-4322-4284-a992-d4296ee90956/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/f66c4453-4322-4284-a992-d4296ee90956/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/f66c4453-4322-4284-a992-d4296ee90956/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/f66c4453-4322-4284-a992-d4296ee90956/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,140 @@
+        }),
+        ref,
+        ...targetEventHandlers,
+    } satisfies React.HTMLProps<HTMLElement>;
```
**note**: Interesting, I haven't seen `satisfies` used. TIL

mm-wang commented on packages/core/src/components/popover-next/popoverPopup.tsx:
```diff
@@ -0,0 +1,164 @@
+        : isClosingViaEscapeKeypress || props.shouldReturnFocusOnClose;
+
+    return (
+        <Overlay2
```
**non-blocking comment**: Would very much love to see the future where we don't use Overlay at all here to minimize this API!

mm-wang commented on packages/core/src/components/popover-next/popoverTarget.tsx:
```diff
@@ -0,0 +1,140 @@
+ * Check if a React element is a Tooltip component by comparing its displayName.
+ * This avoids importing the Tooltip component directly, which would create a dependency cycle.
+ */
+function isTooltipElement(element: React.ReactElement): boolean {
```
**non-blocking comment**: This is interesting, something I'd be keen to remove if we choose to have Tooltip built differently.

ggdouglas replied:
Yes, this was originally split out/added in #7572 to eliminate some of the import cycle issues between Popover and Tooltip. I agree that the coupling would be nice to eventually get rid of.

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+        });
+    });
+
+    describe("key interactions on Button target", () => {
```
**non-blocking comment**: It would be cool to see if we can set up some combination tests, such as Popover with MenuItems, although I think that they are likely better suited to an integration test.

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+        });
+    });
+
+    describe("focus management when shouldReturnFocusOnClose={true}", () => {
```
**non-blocking comment**: I suspect we'll need to opposite test now that `shouldReturnFocusOnClose=
{true}` is the default

ggdouglas replied:
Sure, should be fairly straightforward to add fabfed46388b8d21a77b1e89f06435bab374fd12

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+    });
+
+    describe("basic functionality", () => {
+        it.skip("inherits dark theme from trigger ancestor", () => {
```
**question**: Why is this skipped already?

ggdouglas replied:
Good question. Many of these were skipped for various reasons in the original Popover tests, and I'd carried them over for posterity. However, not that we're using vitest in core, it's much easier/more possible to unskip these. Done in 06774d8eb4b50e70b1b1e9a73c241c496cc4ab63

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+                expect(screen.getByRole("button", { name: "target" }).getAttribute("tabindex")).to.be.null;
+            });
+
+            it.skip("opens popover on target focus when interactionKind is HOVER", async () => {
```
**question**: Why is this skipped already (similar to others)?

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+                expect(targetButton).to.equal(document.activeElement);
+            });
+
+            it.skip("opens popover on target focus when interactionKind is HOVER_TARGET_ONLY", async () => {
```
**question**: Why is this skipped already (similar to others)?

mm-wang replied:
**question**: Why is this skipped already? (Similar to others)

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+            await waitFor(() => expect(screen.queryByRole("button", { name: "close" })).not.to.exist);
+        });
+
+        it.skip("pressing Escape closes popover when canEscapeKeyClose=true and usePortal=false", async () => {
```
**question**: Why is this skipped already (similar to others)?

mm-wang replied:
**question**: Why is this skipped already? (Similar to others)

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+import { PopoverNext } from "../../src/components/popover-next/popoverNext";
+import { hasClass } from "../utils";
+
+describe("<PopoverNext>", () => {
```
**suggestion, non-blocking**: It might be nice to test target + popup behaviors in a different file, but not necessary for this to move forward.

mm-wang replied:
**suggestion, non-blocking**: It might be nice to have some tests that specifically have to do with `popoverPopup` and `popoverTarget` behavior in separate files.

mm-wang commented on packages/core/test/popover-next/popoverNextTests.tsx:
```diff
@@ -0,0 +1,1466 @@
+                expect(screen.getByRole("button", { name: "target" }).getAttribute("tabindex")).to.be.null;
+            });
+
+            it.skip("opens popover on target focus when interactionKind is HOVER", async () => {
```
**question**: Why is this skipped already? (Similar to others)

ggdouglas replied:
These hover interaction tests in particular needed to be modified to reduce flake. There were issues with the assertions and the async behavior of the focus trapping behavior.

mm-wang reviewed (dismissed):
A saga! Looks good, I have some nonblocking comments and notes for the future. In general I think this is fine to put in and to start decomping the next components to see what changes happen.

Chatted about the use of ResizeObserver not necessary here as Floating UI should support those interactions now.

ggdouglas committed:
Merge branch 'develop' into gdouglas/popover-next

svc-palantir-github commented:
<h3>Merge branch 'develop' into gdouglas/popover-next</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/4261d8b5-750b-4a33-bd87-dc11bd212e34/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/4261d8b5-750b-4a33-bd87-dc11bd212e34/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/4261d8b5-750b-4a33-bd87-dc11bd212e34/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/4261d8b5-750b-4a33-bd87-dc11bd212e34/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Merge branch 'develop' into gdouglas/popover-next

ggdouglas committed:
Update PopoverNext tests to be compatible with vitest framework

ggdouglas committed:
Unskip a bunch of tests in PopoverNext

ggdouglas committed:
Add mirror test cases for `shouldReturnFocusOnClose`

svc-palantir-github commented:
<h3>Add mirror test cases for `shouldReturnFocusOnClose`</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/61d57779-5487-4b0e-b14a-844676a7083a/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/61d57779-5487-4b0e-b14a-844676a7083a/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/61d57779-5487-4b0e-b14a-844676a7083a/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/61d57779-5487-4b0e-b14a-844676a7083a/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>
